staging/src/k8s.io/kubectl/pkg/cmd/expose/expose.go

```diff
@@ -165,56 +181,75 @@ func NewCmdExposeService(f cmdutil.Factory, streams genericiooptions.IOStreams)
+	cmd.Flags().StringVar(&flags.exposeInputFlags.Selector, "selector", flags.exposeInputFlags.Selector, i18n.T("A label selector to use for this service. Only equality-based selector requirements are supported. If empty (the default) infer the selector from the replication controller or replica set.)"))
+	cmd.Flags().StringVarP(&flags.exposeInputFlags.Labels, "labels", "l", flags.exposeInputFlags.Labels, "Labels to apply to the service created by this call.")
+	cmd.Flags().StringVar(&flags.exposeInputFlags.TargetPort, "target-port", flags.exposeInputFlags.TargetPort, i18n.T("Name or number for the port on the container that the service should direct traffic to. Optional."))
+	cmd.Flags().StringVar(&flags.exposeInputFlags.ExternalIP, "external-ip", flags.exposeInputFlags.ExternalIP, i18n.T("Additional external IP address (not managed by Kubernetes) to accept for the service. If this IP is routed to a node, the service can be accessed by this IP in addition to its generated service IP."))
+	cmd.Flags().StringVar(&flags.exposeInputFlags.Name, "name", flags.exposeInputFlags.Name, i18n.T("The name for the newly created object."))
+	cmd.Flags().StringVar(&flags.exposeInputFlags.SessionAffinity, "session-affinity", flags.exposeInputFlags.SessionAffinity, i18n.T("If non-empty, set the session affinity for the service to this; legal values: 'None', 'ClientIP'"))
+	cmd.Flags().StringVar(&flags.exposeInputFlags.ClusterIP, "cluster-ip", flags.exposeInputFlags.ClusterIP, i18n.T("ClusterIP to be assigned to the service. Leave empty to auto-allocate, or set to 'None' to create a headless service."))
+
+	cmdutil.AddFieldManagerFlagVar(cmd, &flags.fieldManager, "kubectl-expose")
+	flags.exposeInputFlags.AddOverrideFlags(cmd)
+
+	usage := "identifying the resource to expose a service"
+	cmdutil.AddFilenameOptionFlags(cmd, &flags.exposeInputFlags.FilenameOptions, usage)
+}
+
+func (flags *exposeServiceFlags) ToOptions(cmd *cobra.Command, args []string) (*ExposeServiceOptions, error) {
+	dryRunStratergy, err := cmdutil.GetDryRunStrategy(cmd)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	cmdutil.PrintFlagsWithDryRunStrategy(o.PrintFlags, o.DryRunStrategy)
-	printer, err := o.PrintFlags.ToPrinter()
+	cmdutil.PrintFlagsWithDryRunStrategy(flags.printFlags, dryRunStratergy)
+	printer, err := flags.printFlags.ToPrinter()
 	if err != nil {
-		return err
+		return nil, err
 	}
-	o.PrintObj = printer.PrintObj
 
-	o.RecordFlags.Complete(cmd)
-	o.Recorder, err = o.RecordFlags.ToRecorder()
+	flags.recordFlags.Complete(cmd)
+	recorder, err := flags.recordFlags.ToRecorder()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	e := &ExposeServiceOptions{
+		DryRunStrategy:     dryRunStratergy,
+		PrintObj:           printer.PrintObj,
+		Recorder:           recorder,
+		ExposeInputOptions: flags.exposeInputFlags,
```

Comment:
Would be better to set options directly

Reply:
@ardaguclu I did this intentionally so that we need not repeat the same fields in both the places when its a direct translation. Plus accepting options and flags at the same time with cmd flags may turn out confusing. 

I'm fine either ways, as you mentioned if making them separate is better I can refactor.

Reply:
Yes, agree that this causes repetition but our main motivation is to decouple them by accepting some repetition and flags will contain flag definitions and options will be converted via flags.